plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSink.java

```diff
@@ -326,14 +329,24 @@ private WriteContext createWriter(Optional<PartitionData> partitionData)
         Object[] values = new Object[columns.size()];
         for (int i = 0; i < columns.size(); i++) {
             PartitionColumn column = columns.get(i);
-            Block block = page.getBlock(column.getSourceChannel());
+            Block block = getPartitionBlock(column, page);
             Type type = column.getSourceType();
             Object value = getIcebergValue(block, position, type);
             values[i] = applyTransform(column.getField().transform(), value);
         }
         return Optional.of(new PartitionData(values));
     }
 
+    private static Block getPartitionBlock(PartitionColumn column, Page page)
+    {
+        Iterator<Integer> pos = column.getSourceChannels().listIterator();
+        Block block = page.getBlock(pos.next());
+        while (pos.hasNext()) {
+            block = block.getChildren().get(pos.next());
```

Comment:
in https://github.com/trinodb/trino/pull/9354 @djsstarburst argues that using `getChildren` is incorrect.

@djsstarburst would you want to comment here?